// @redux-devtools/log-monitor/lib/cjs/LogMonitorEntry.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _reactJsonTree = require("react-json-tree");
var _LogMonitorEntryAction = _interopRequireDefault(require("./LogMonitorEntryAction"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const styles = {
  entry: {
    display: 'block',
    WebkitUserSelect: 'none'
  },
  root: {
    marginLeft: 0
  }
};
const getDeepItem = (data, path) => path.reduce((obj, key) => obj && obj[key], data);
const dataIsEqual = (data, previousData, keyPath) => {
  const path = [...keyPath].reverse().slice(1);
  return getDeepItem(data, path) === getDeepItem(previousData, path);
};
class LogMonitorEntry extends _react.PureComponent {
  constructor() {
    super(...arguments);
    (0, _defineProperty2.default)(this, "handleActionClick", e => {
      const {
        actionId,
        onActionClick,
        onActionShiftClick
      } = this.props;
      if (actionId > 0) {
        if (e.shiftKey) {
          onActionShiftClick(actionId);
        } else {
          onActionClick(actionId);
        }
      }
    });
    (0, _defineProperty2.default)(this, "shouldExpandNodeInitially", (keyPath, data, level) => {
      return this.props.expandStateRoot && level === 0;
    });
  }
  printState(state, error) {
    let errorText = error;
    if (!errorText) {
      try {
        const data = this.props.select(state);
        let theme;
        if (this.props.markStateDiff) {
          const previousData = typeof this.props.previousState !== 'undefined' ? this.props.select(this.props.previousState) : undefined;
          const getValueStyle = (_ref, nodeType, keyPath) => {
            let {
              style
            } = _ref;
            return {
              style: {
                ...style,
                backgroundColor: dataIsEqual(data, previousData, keyPath) ? 'transparent' : this.props.theme.base01
              }
            };
          };
          const getNestedNodeStyle = (_ref2, keyPath) => {
            let {
              style
            } = _ref2;
            return {
              style: {
                ...style,
                ...(keyPath.length > 1 ? {} : styles.root)
              }
            };
          };
          theme = {
            extend: this.props.theme,
            value: getValueStyle,
            nestedNode: getNestedNodeStyle
          };
        } else {
          theme = this.props.theme;
        }
        return /*#__PURE__*/_react.default.createElement(_reactJsonTree.JSONTree, {
          theme: theme,
          data: data,
          invertTheme: false,
          keyPath: ['state'],
          shouldExpandNodeInitially: this.shouldExpandNodeInitially
        });
      } catch (err) {
        errorText = 'Error selecting state.';
      }
    }
    return /*#__PURE__*/_react.default.createElement("div", {
      style: {
        color: this.props.theme.base08,
        paddingTop: 20,
        paddingLeft: 30,
        paddingRight: 30,
        paddingBottom: 35
      }
    }, errorText);
  }
  render() {
    const {
      actionId,
      error,
      action,
      state,
      collapsed,
      selected,
      inFuture
    } = this.props;
    const styleEntry = {
      opacity: collapsed ? 0.5 : 1,
      cursor: actionId > 0 ? 'pointer' : 'default'
    };
    return /*#__PURE__*/_react.default.createElement("div", {
      style: {
        opacity: selected ? 0.4 : inFuture ? 0.6 : 1,
        // eslint-disable-line no-nested-ternary
        textDecoration: collapsed ? 'line-through' : 'none',
        color: this.props.theme.base06
      }
    }, /*#__PURE__*/_react.default.createElement(_LogMonitorEntryAction.default, {
      theme: this.props.theme,
      collapsed: collapsed,
      action: action,
      expandActionRoot: this.props.expandActionRoot,
      onClick: this.handleActionClick,
      style: {
        ...styles.entry,
        ...styleEntry
      }
    }), !collapsed && /*#__PURE__*/_react.default.createElement("div", {
      style: {
        paddingLeft: 16
      }
    }, this.printState(state, error)));
  }
}
exports.default = LogMonitorEntry;
(0, _defineProperty2.default)(LogMonitorEntry, "propTypes", {
  state: _propTypes.default.object.isRequired,
  action: _propTypes.default.object.isRequired,
  actionId: _propTypes.default.number.isRequired,
  select: _propTypes.default.func.isRequired,
  inFuture: _propTypes.default.bool,
  error: _propTypes.default.string,
  onActionClick: _propTypes.default.func.isRequired,
  onActionShiftClick: _propTypes.default.func.isRequired,
  collapsed: _propTypes.default.bool,
  selected: _propTypes.default.bool,
  expandActionRoot: _propTypes.default.bool,
  expandStateRoot: _propTypes.default.bool,
  previousState: _propTypes.default.object
});
