// @redux-devtools/log-monitor/lib/cjs/LogMonitorEntryAction.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _reactJsonTree = require("react-json-tree");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const styles = {
  actionBar: {
    paddingTop: 8,
    paddingBottom: 7,
    paddingLeft: 16
  },
  payload: {
    margin: 0,
    paddingLeft: 16,
    overflow: 'auto'
  }
};
class LogMonitorAction extends _react.Component {
  constructor() {
    super(...arguments);
    (0, _defineProperty2.default)(this, "shouldExpandNodeInitially", (keyPath, data, level) => {
      return this.props.expandActionRoot && level === 0;
    });
  }
  renderPayload(payload) {
    return /*#__PURE__*/_react.default.createElement("div", {
      style: {
        ...styles.payload,
        backgroundColor: this.props.theme.base00
      }
    }, Object.keys(payload).length > 0 ? /*#__PURE__*/_react.default.createElement(_reactJsonTree.JSONTree, {
      theme: this.props.theme,
      invertTheme: false,
      keyPath: ['action'],
      data: payload,
      shouldExpandNodeInitially: this.shouldExpandNodeInitially
    }) : '');
  }
  render() {
    const {
      type,
      ...payload
    } = this.props.action;
    return /*#__PURE__*/_react.default.createElement("div", {
      style: {
        backgroundColor: this.props.theme.base02,
        color: this.props.theme.base06,
        ...this.props.style
      }
    }, /*#__PURE__*/_react.default.createElement("div", {
      style: styles.actionBar,
      onClick: this.props.onClick
    }, type !== null && type.toString()), !this.props.collapsed ? this.renderPayload(payload) : '');
  }
}
exports.default = LogMonitorAction;
